### PR TITLE
Per-object variable precision + tau ancestor dressed lepton bit

### PR DESCRIPTION
**Pull request by emanca**
as discussed with @peruzzim and @intrepid42

### CODE REVIEW

**Comment by arizzi** [comment]
what is W mass specific to have W mass in the name?
I understood the point was to ID dressed muon that originates from tau, it doesn't seem very WMass specific to me...

**Comment by emanca** [comment]
> what is W mass specific to have W mass in the name?
> I understood the point was to ID dressed muon that originates from tau, it doesn't seem very WMass specific to me...

In the end we have modified the whole collection of GenDressedLeptons. The collection with Wmass in the name is a copy of the charged leptons and neutrinos from the GenParticles with increased numerical precision. Call it the way you prefer :D

File: PhysicsTools/NanoAOD/python/genparticles_cff.py [hunk 1]
```diff
@@ -81,7 +81,68 @@
...
+    doc = cms.string("Charged leptons and neutrinos from GenParticles at high numerical precision"),
+    singleton = cms.bool(False), # the number of entries is variable
+    extension = cms.bool(False), # this is the main table for the taus
+    variables = cms.PSet(
+         pt  = Var("pt",  float),
+         phi = Var("phi", float),
+         eta  = Var("eta",  float),
+         pdgId  = Var("pdgId", int, doc="PDG id"),
+         status  = Var("status", int, doc="Particle status. 1=stable"),
+         genPartIdxMother = Var("?numberOfMothers>0?motherRef(0).key():-1", int, doc="index of the mother particle"),
```
Comment (arizzi) [inline PhysicsTools/NanoAOD/python/genparticles_cff.py]: this is the wrong name it should reflect the name of the pointed collection

**Comment by arizzi** [comment]
try to give descriptive names, Wmass is not.

anyhow what is the size impact of repeating some genparticles multiple time wrt just increasing the size as needed for all particles?

what is the numerical precision that you need?

in principle we can increase the numerical precision only for individual particles in a collection (or actually the other way, we can reduce the precision only for some)

**Comment by mseidel42** [comment]
Hi, indeed it would be nice to have top quarks stored at a higher precision, too. The current precision does not allow for reweighting the top width I think.

Also, the W bosons themselves are needed in good precision for doing W mass reweighting, for sure!

**Comment by emanca** [comment]
@intrepid42 for w mass we don’t use W bosons from the generator but we rather reconstruct its 4-vector using the charged lepton and neutrino. 
@arizzi we didn’t make any memory test (I know that gen particles are a pain here memory-wise) so if you and @peruzzim have some smarter idea than duplicate the collection we can implement this.

**Comment by mseidel42** [comment]
> @intrepid42 for w mass we don’t use W bosons from the generator but we rather reconstruct its 4-vector using the charged lepton and neutrino.

Hm, that could fail once we use Photos because it may insert 1->3 vertices W>lepton+neutrino+photon. Better to include the status 62 W boson with good precision!

File: PhysicsTools/NanoAOD/python/particlelevel_cff.py [hunk 2]
```diff
@@ -46,6 +46,10 @@
     fatJetMaxEta   = cms.double(999.),
 )
 
+tautagger = cms.EDProducer("GenJetTauTaggerProducer",
+    src = cms.InputTag("rivetLeptonTable.src"),
```
Comment (peruzzim) [inline PhysicsTools/NanoAOD/python/particlelevel_cff.py]: to use rivetLeptonTable.src it should not be inside cms.InputTag

File: PhysicsTools/NanoAOD/python/particlelevel_cff.py [hunk 3]
```diff
@@ -61,6 +65,9 @@
     doc = cms.string("Dressed leptons from Rivet-based ParticleLevelProducer"),
     singleton = cms.bool(False), # the number of entries is variable
     extension = cms.bool(False), # this is the main table
+    externalVariables = cms.PSet(
+        tautagger = ExtVar(cms.InputTag("tautagger"),bool, doc="true if Dressed lepton has a tau as ancestor"),
```
Comment (peruzzim) [inline PhysicsTools/NanoAOD/python/particlelevel_cff.py]: please name this `hasTauAnc`

File: PhysicsTools/NanoAOD/python/genparticles_cff.py (hunk 1 above)
Comment (peruzzim) [reply PhysicsTools/NanoAOD/python/genparticles_cff.py]: please follow up on the comment above

File: PhysicsTools/NanoAOD/python/genparticles_cff.py [hunk 4]
```diff
@@ -81,7 +81,68 @@
...
+##################### Tables for final output and docs ##########################
+genParticleTableWmass = cms.EDProducer("SimpleCandidateFlatTableProducer",
+    src = cms.InputTag("finalGenParticlesWmass"),
+    cut = cms.string(""), #we should not filter after pruning
+    name= cms.string("GenPartWmass"),
+    doc = cms.string("Charged leptons and neutrinos from GenParticles at high numerical precision"),
+    singleton = cms.bool(False), # the number of entries is variable
+    extension = cms.bool(False), # this is the main table for the taus
+    variables = cms.PSet(
+         pt  = Var("pt",  float),
```
Comment (peruzzim) [inline PhysicsTools/NanoAOD/python/genparticles_cff.py]: precision specifiers are missing for all these variables, but we should move to the model of per-particle precision enabled by the new feature, right? so I guess this table is all to be dropped and the per-particle precision be defined in the other table according to your needs

File: PhysicsTools/NanoAOD/python/particlelevel_cff.py [hunk 5]
```diff
@@ -31,11 +31,11 @@
...
     
     particleMinPt  = cms.double(0.),
-    particleMaxEta = cms.double(5.), # HF range. Maximum 6.0 on MiniAOD
+    particleMaxEta = cms.double(6.), # HF range. Maximum 6.0 on MiniAOD
     
     lepConeSize = cms.double(0.1), # for photon dressing
-    lepMinPt    = cms.double(15.),
-    lepMaxEta   = cms.double(2.5),
+    lepMinPt    = cms.double(0.),
+    lepMaxEta   = cms.double(10.0),
```
Comment (peruzzim) [inline PhysicsTools/NanoAOD/python/particlelevel_cff.py]: really lepMaxEta = 10, while particleMaxEta = 6? could you please cross-check?

**Review by peruzzim — DISMISSED** [review]

**Comment by peruzzim** [comment]
As it is, I'm afraid this PR cannot go in without further studies and tailoring of the needed information:

- the size increase is very large, mainly driven by the additional precision in the GenPart collection: +130 B/ev. in ttbar, then another +25 B/ev. in GenDressedLepton because of its much larger multiplicity due to much looser acceptance cuts
- the fiducialMet variables are altered by the custom parameters passed to ParticleLevelProducer
- the analysis would probably need to increase the precision on some reconstructed lepton variables (currently: pt=23, eta=12, phi=12, mass=10) to really profit from this

**Comment by peruzzim** [comment]
@emanca @intrepid42 the validation shows that the newly added variable `hasTauAnc` is always false in ttbar. Could you please comment whether this is expected (I think not), and check the code?

**Comment by peruzzim** [comment]
> @emanca @intrepid42 the validation shows that the newly added variable `hasTauAnc` is always false in ttbar. Could you please comment whether this is expected (I think not), and check the code?

I merge the PR to have the configurable precision functionality, but please be sure to check the above. Thanks

**Pull request merged**